cumulus/pallets/xcmp-queue/src/migration/v5.rs

```diff
@@ -0,0 +1,119 @@
+	type ChannelList: ListChannelInfos;
+}
+
+/// Ensures that the storage migrates cleanly to V5.
+///
+/// The migration itself is a no-op, but it checks that none of the `BoundedVec`s would truncate on
+/// the next decode after the upgrade was applied.
+pub type MigrateV4ToV5<T> = frame_support::migrations::VersionedMigration<
+	4,
+	5,
+	unversioned::UncheckedMigrateV4ToV5<T>,
+	Pallet<T>,
+	<T as frame_system::Config>::DbWeight,
+>;
+
+// V4 storage aliases
+mod v4 {
+	use super::*;
+
+	#[frame_support::storage_alias]
+	pub(super) type OutboundXcmpStatus<T: Config> =
+		StorageValue<Pallet<T>, Vec<OutboundChannelDetails>, ValueQuery>;
+
+	#[frame_support::storage_alias]
+	pub(super) type OutboundXcmpMessages<T: Config> = StorageDoubleMap<
+		Pallet<T>,
+		Blake2_128Concat,
+		ParaId,
+		Twox64Concat,
+		u16,
+		Vec<u8>,
+		ValueQuery,
+	>;
+
+	#[frame_support::storage_alias]
+	pub(super) type SignalMessages<T: Config> =
+		StorageMap<Pallet<T>, Blake2_128Concat, ParaId, Vec<u8>, ValueQuery>;
+}
+
+// Private module to hide the migration.
+mod unversioned {
+	/// Please use [`MigrateV4ToV5`] instead.
+	pub struct UncheckedMigrateV4ToV5<T: super::V5Config>(core::marker::PhantomData<T>);
+}
+
+impl<T: V5Config> UncheckedOnRuntimeUpgrade for unversioned::UncheckedMigrateV4ToV5<T> {
+	fn on_runtime_upgrade() -> frame_support::weights::Weight {
+		Default::default()
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade(_: Vec<u8>) -> Result<(), sp_runtime::DispatchError> {
+		// We dont need any front-run protection for this since channels are opened by governance.
+		ensure!(
+			v4::OutboundXcmpStatus::<T>::get().len() as u32 <= T::MaxActiveOutboundChannels::get(),
+			"Too many outbound channels. Close some channels or increase `MaxActiveOutboundChannels`."
+		);
+
+		// Check if any channels have a too large message max sizes.
+		let max_msg_len = T::MaxPageSize::get() - XcmpMessageFormat::max_encoded_len() as u32;
+		for channel in T::ChannelList::outgoing_channels() {
+			let info = T::ChannelInfo::get_channel_info(channel)
+				.expect("All listed channels must provide info");
+
+			if info.max_message_size > max_msg_len {
+				panic!(
+					"Max message size for channel is too large. This means that the V5 \
+				migration can be front-run and an attacker could place a large message just right \
+				before the migration to make other messages un-decodable. Please either increase \
+				`MaxPageSize` or decrease the `max_message_size` for this channel. Channel max: {}, \
+				MaxPageSize: {}",
+					info.max_message_size, max_msg_len
+				);
+				//return Err("Migration can be front-run".into());
+			}
+		}
+
+		// Now check that all pages still fit into the new `BoundedVec`s:
+		for page in v4::OutboundXcmpMessages::<T>::iter_values() {
+			if page.len() >= T::MaxPageSize::get() as usize {
+				log::warn!("Too long message in storage. WeakBoundedVec should handle this.");
```

Comment:
How does `WeakBoundedVec` handle this?

Reply:
It just truncates 😬 will change comment.